storage/database/dynamodb.go

```diff
@@ -0,0 +1,487 @@
+				B: key,
+			},
+		},
+		ConsistentRead: aws.Bool(true),
+	}
+
+	result, err := dynamo.db.GetItem(params)
+	if err != nil {
+		fmt.Printf("Get ERROR: %v\n", err.Error())
+		return nil, err
+	}
+
+	if result.Item == nil {
+		return nil, dataNotFoundErr
+	}
+
+	var data DynamoData
+	if err := dynamodbattribute.UnmarshalMap(result.Item, &data); err != nil {
+		return nil, err
+	}
+
+	if data.Val == nil {
+		return []byte{}, nil
+	}
+
+	if bytes.Equal(data.Val, overSizedDataPrefix) {
+		return dynamo.fdb.read(key)
+	}
+
+	return data.Val, nil
+}
+
+// Delete deletes the key from the queue and database
+func (dynamo *dynamoDB) Delete(key []byte) error {
+	params := &dynamodb.DeleteItemInput{
+		TableName: aws.String(dynamo.config.TableName),
+		Key: map[string]*dynamodb.AttributeValue{
+			"Key": {
+				B: key,
+			},
+		},
+	}
+
+	_, err := dynamo.db.DeleteItem(params)
+	if err != nil {
+		fmt.Printf("ERROR: %v\n", err.Error())
+		return err
+	}
+	return nil
+}
+
+func (dynamo *dynamoDB) Close() {
+	close(dynamo.quitCh)
+}
+
+func (dynamo *dynamoDB) Meter(prefix string) {
+	// TODO-Klaytn: implement this later.
+	dynamo.batchWriteTimeMeter = metrics.NewRegisteredMeter(prefix+"batchwrite/time", nil)
+	dynamo.batchWriteCountMeter = metrics.NewRegisteredMeter(prefix+"batchwrite/count", nil)
+	dynamo.batchWriteSizeMeter = metrics.NewRegisteredMeter(prefix+"batchwrite/size", nil)
+	dynamo.batchWriteSecPerItemMeter = metrics.NewRegisteredMeter(prefix+"batchwrite/secperitem", nil)
+	dynamo.batchWriteSecPerByteMeter = metrics.NewRegisteredMeter(prefix+"batchwrite/secperbyte", nil)
+}
+
+func (dynamo *dynamoDB) NewIterator() Iterator {
+	// TODO-Klaytn: implement this later.
+	return nil
+}
+
+func (dynamo *dynamoDB) NewIteratorWithStart(start []byte) Iterator {
+	// TODO-Klaytn: implement this later.
+	return nil
+}
+
+func (dynamo *dynamoDB) NewIteratorWithPrefix(prefix []byte) Iterator {
+	// TODO-Klaytn: implement this later.
+	return nil
+}
+
+func (dynamo *dynamoDB) createBatchWriteWorker() {
+	failCount := 0
+	tableName := dynamo.config.TableName
+	batchWriteInput := &dynamodb.BatchWriteItemInput{
+		RequestItems: map[string][]*dynamodb.WriteRequest{},
+	}
+	dynamo.logger.Info("generate a dynamoDB batchWrite worker")
+
+	for {
+		select {
+		case <-dynamo.quitCh:
+			dynamo.logger.Info("close a dynamoDB batchWrite worker")
+			return
+		case batchInput := <-dynamo.writeCh:
+			batchWriteInput.RequestItems[tableName] = batchInput.items
+
+			BatchWriteItemOutput, err := dynamo.db.BatchWriteItem(batchWriteInput)
+			numUnprocessed := len(BatchWriteItemOutput.UnprocessedItems[tableName])
+			for err != nil || numUnprocessed != 0 {
+				if err != nil {
+					failCount++
+					dynamo.logger.Warn("dynamoDB failed to write batch items", "err", err, "failCnt", failCount)
+					if failCount > dynamoMaxRetry {
+						dynamo.logger.Error("dynamoDB failed many times. sleep a second and retry",
+							"failCnt", failCount)
+						time.Sleep(time.Second)
```

Comment:
I don't know why this sleep is need.

And we need to consider how to deal with the infinite fail case.

Reply:
The DynamoDB document says it is a best practice to have a retry logic with delay.
The sleeping only happens when the DynamoDB is failing to put items more than 5 times.
This might not be the best retry logic, it seems okay for me.